src/directory/nrt_directory.rs

```diff
@@ -0,0 +1,141 @@
+use std::collections::HashSet;
+use std::fmt;
+use std::io::{self, Write};
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, RwLock};
+
+use crate::directory::error::{DeleteError, OpenReadError, OpenWriteError};
+use crate::directory::{Directory, DirectoryLock, FileHandle, FileSlice, Lock, TerminatingWrite,
+                       WatchCallback, WatchHandle, WritePtr};
+use crate::directory::RamDirectory;
+
+/// A Directory that overlays a `RamDirectory` on top of a base `Directory`.
+///
+/// - Writes (open_write and atomic_write) go to the in-memory overlay.
+/// - Reads first check the overlay, then fallback to the base directory.
+/// - sync_directory() persists overlay files that do not yet exist in the base directory.
+#[derive(Clone)]
+pub struct NrtDirectory {
+    base: Box<dyn Directory>,
+    overlay: RamDirectory,
+    /// Tracks files written into the overlay to decide what to persist on sync.
+    overlay_paths: Arc<RwLock<HashSet<PathBuf>>>,
+}
+
+impl NrtDirectory {
+    /// Wraps a base directory with an NRT overlay.
+    pub fn wrap(base: Box<dyn Directory>) -> NrtDirectory {
+        NrtDirectory {
+            base,
+            overlay: RamDirectory::default(),
+            overlay_paths: Arc::new(RwLock::new(HashSet::new())),
+        }
+    }
+
+    /// Persist overlay files into the base directory if missing there.
+    fn persist_overlay_into_base(&self) -> crate::Result<()> {
+        let snapshot_paths: Vec<PathBuf> = {
+            let guard = self.overlay_paths.read().unwrap();
+            guard.iter().cloned().collect()
+        };
+        for path in snapshot_paths {
```

Comment:
ok, but if we update the meta json file before a segment file, and this function has an io Error in the middle, the index will be corrupted right now?